x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java

```diff
@@ -0,0 +1,376 @@
+                        context.addUnsuccessfulMessage(authenticator.name() + ": " + e.getMessage());
+                        listener.onResponse(Authenticator.Result.unsuccessful(e.getMessage(), e));
+                    }
+                    return;
+                }
+                if (authenticationToken == null) {
+                    listener.onResponse(Authenticator.Result.notHandled());
+                    return;
+                }
+                context.addAuthenticationToken(authenticationToken);
+            }
+            context.setHandleNullToken(context.shouldHandleNullToken() && authenticator.canFollowedByNullTokenHandler());
+            authenticator.authenticate(context, ActionListener.wrap(result -> {
+                if (result.getStatus() == Authenticator.Status.UNSUCCESSFUL) {
+                    context.addUnsuccessfulMessage(authenticator.name() + ": " + result.getMessage());
+                }
+                listener.onResponse(result);
+            }, e -> {
+                if (e instanceof ElasticsearchSecurityException) {
+                    final ElasticsearchSecurityException ese = (ElasticsearchSecurityException) e;
+                    if (false == context.getUnsuccessfulMessages().isEmpty()) {
+                        addMetadata(context, ese);
+                    }
+                }
+                // Not adding additional metadata if the exception is not security related, e.g. server busy
+                listener.onFailure(e);
+            }));
+        };
+    }
+
+    private void maybeLookupRunAsUser(
+        Authenticator.Context context,
+        Authentication authentication,
+        ActionListener<Authentication> listener
+    ) {
+        // TODO: only allow run as for realm authentication to maintain the existing behaviour
+        if (false == runAsEnabled || authentication.getAuthenticationType() != Authentication.AuthenticationType.REALM) {
+            finishAuthentication(context, authentication, listener);
+            return;
+        }
+
+        final String runAsUsername = context.getThreadContext().getHeader(AuthenticationServiceField.RUN_AS_USER_HEADER);
+        if (runAsUsername == null) {
+            finishAuthentication(context, authentication, listener);
+            return;
+        }
+
+        final User user = authentication.getUser();
+        if (runAsUsername.isEmpty()) {
+            logger.debug("user [{}] attempted to runAs with an empty username", user.principal());
+            listener.onFailure(context.getRequest()
+                .runAsDenied(new Authentication(new User(runAsUsername, null, user),
+                    authentication.getAuthenticatedBy(),
+                    null), context.getAuthenticationToken()));
+            return;
+        }
+
+        // Now we have a valid runAsUsername
+        realmsAuthenticator.lookupRunAsUser(context, authentication, ActionListener.wrap(tuple -> {
```

Comment:
That is an interesting idea. By "pass it through the whole chain again", do you mean reuse the same `authenticateAsyncWithExistingCredentials` method? Orr do you mean resuse just the whole chain but have a different entry method? I

I think reuse the whole chain but with a different entry method would be better since run-as does not verify credentials and also does not handle null token etc. For example, we could extend the `Authenticator` interface to support `lookupRunAsUser` the same way as the `RealmsAuthenticator` does. 

That said, currently no meaningful run-as support can be defined for authenticators other than Realms. Run as API key or token does not seem to make sense. For service token, we decided it should *not* be the target of run-as. Therefore, I think we can revisit it later.